Wrap C10 CUDAStream instead of cudaStream_t in THCPStream

mrshenli opened the pull request.

mrshenli committed:
Wrap C10 CUDAStream in THCPStream instead of cudaStream_t

mrshenli commented on torch/csrc/cuda/Stream.h:
```diff
@@ -11,8 +12,7 @@ struct THCPStream {
   // the hook for placement-new'ing/delete'ing it; simpler to just
```
Doesn't the comment above apply to `CUDAStream` as well? It is not a POD either.

colesbury replied:
This code comment is no longer relevant (and should be deleted)

mrshenli commented on torch/csrc/cuda/Stream.h:
```diff
@@ -11,8 +12,7 @@ struct THCPStream {
   // the hook for placement-new'ing/delete'ing it; simpler to just
   // rely on the packed representation...)
   uint64_t cdata;
-  int device;
```
Removing these two member vars changes the behavior of accessing `device` and `cuda_stream`. Although these two members are not advertised in our doc, will this be a problem for our users? Or should I keep them in `PyMemberDef` and just point them to new locations?

mrshenli replied:
I think I should keep those two vars.

mrshenli reviewed (commented):
@colesbury

mrshenli committed:
add device and cuda_stream properties to Stream for back compatibility

mrshenli committed:
revert changes of moving __hash__

colesbury commented on torch/csrc/cuda/Stream.cpp:
```diff
@@ -38,27 +37,46 @@ static PyObject * THCPStream_pynew(PyTypeObject *type, PyObject *args, PyObject
+
+static PyObject * THCPStream_device(THCPStream *self) {
+  HANDLE_TH_ERRORS
+  return PyLong_FromLong(self->cuda_stream.device_index());
```
Prefer `THPUtils_packInt64` from `python_numbers.h`. It has slightly nicer behavior in Python 2 (preferring `int` when the value is small enough) .

mrshenli commented:
Thanks! That's very thorough.

colesbury commented on torch/csrc/cuda/Stream.cpp:
```diff
@@ -38,27 +37,46 @@ static PyObject * THCPStream_pynew(PyTypeObject *type, PyObject *args, PyObject
+
+static PyObject * THCPStream_stream(THCPStream *self) {
+  HANDLE_TH_ERRORS
+  return PyLong_FromUnsignedLongLong(
```
I think you should try to keep the same property names in Python (i.e. `stream.cuda_stream` here instead of `stream.stream()`)

cuda_stream.stream() returns a `cudaStream_t`, which is a pointer type so you should probably use `PyLong_FromVoidPtr` instead.

mrshenli committed:
Address comments

1. expose device and cuda_stream properties through PyGetSetDev
2. use placement new and delete to create and free CUDAStream obj

mrshenli committed:
add tests for exposed properties

mrshenli committed:
remove obsolete comments

mrshenli committed:
Merge remote-tracking branch 'upstream/master' into cudastream
